Backup options

murfdog64 opened the pull request:
Re-enabled backup functionality, up to choosing backup location and the display of days/number of backup (will start on full functionality for that next![image](https://user-images.githubusercontent.com/61074332/117025688-5c740780-acb8-11eb-86e6-304a0c89dc25.png)

murfdog64 committed:
Enable backup options

build actual backup location logic

adjusted page location, removed unnecessary state

adjusted defaultValue for select

changed backup input to uncontrolled input

fixed empty or zero input error

added seperate input for each backup type

Add a backup options component to encapsulate the change logic

jeanawhou commented on src/dashboard/components/options/OptionsHome.js:
```diff
@@ -139,34 +141,37 @@ export default function OptionsHome(props) {
+          <h4>{t('Auto-Save')}</h4>
+          <Switch
+            disabled
+            isOn={!!settings.user.autoSave || true}
```
I think `settings.user.autoSave` will be enough

murfdog64 replied:
I didn't actually write any of that section, it moved down when I re-enabled the other parts. That's still waiting for activation another day looks like :)

Quiescent commented on src/dashboard/components/options/OptionsHome.js:
```diff
@@ -35,24 +37,24 @@ export default function OptionsHome(props) {
+      setBackupLocation(folderPath)
+      saveSetting('user.backupLocation', folderPath)
+    }
+    setLoading(false)
```
Why do we unset loading and pre-loading here?

Quiescent replied:
I spoke to @dstonem about `loading` and `preloading`, and he said that it was part of another attempt at this feature.
I'd recommend removing lines 21-38 (from where `loading` is defined up until the `useEffect`, which uses them) and then removing the `blur` and `focus` handlers in the `div` at the root of this component's render tree.

Quiescent commented:
Are these beta features?
(I'm not sure that they are because that would imply that the user could opt-out or that they're unstable.
It seems that these features are on and mandatory and that the user has no choice in using them.)

Quiescent commented on src/dashboard/components/options/OptionsHome.js:
```diff
@@ -35,24 +37,24 @@ export default function OptionsHome(props) {
+    const files = dialog.showOpenDialogSync(win, { title, properties })
+    if (files && files.length) {
+      let folderPath = files[0]
+      setBackupLocation(folderPath)
```
I'm not sure why we have both state and a setting for the backup location.
Looking at the way that the state for backup location works, we're setting the state inside of the render function(!)
I'm not sure that it's necessary to have state at all here, but if we need it then we should use the `useEffect` hook to trigger changes to it either on the first component mount or when some prop changes.

Quiescent commented on src/dashboard/components/options/OptionsHome.js:
```diff
@@ -139,34 +141,37 @@ export default function OptionsHome(props) {
           <h4>{t('Days of Backup')}</h4>
           <FormGroup controlId="backupDays">
-            <select onChange={(event) => setBackupType(event.target.value)} disabled>
+            <select onChange={(event) => setBackupType(event.target.value)}>
```
This state is internal to the component.  i.e. if we refresh the page, the choice reverts to its default(!)
Again, I'm not sure that the state's necessary here.
We should use and alter the value in the settings store directly.

murfdog64 commented:
https://user-images.githubusercontent.com/61074332/117364343-03a09c80-ae7b-11eb-8fd0-4abb8364e021.mov

forgot the movie

Quiescent commented:
Looks great and works well.  One problem: the option in the `<select>` isn't set when the page initially loads.
You can see the problem if you set it to "Number of backups" and then refresh the dashboard.
It'll show the correct value for the number of backups (and it probably has the right type of backup in the settings) but the select spuriously says "Days of backups" instead of "Number of backups".
The solution is to give the `select` a `defaultValue` or a `value` which is the value of the setting.

Quiescent commented:
I just tested this on Windows and found another three problems.
(Better now than when they land on prod! :) )

1. there's no default value for `backupDays`, so when you select backup days and don't change the value, and then come back, you'll find it empty.  It's only possible to test this on a fresh install, so if you want to replicate it, then be sure to delete your config files in Plottr's app data directory (I can't remember off the top of my head where that is in macOS).
2. if you try to backspace all of the input, it stops at 0.  (I'm guilty of a similar mistake in the Act Structure Modal).  The solution is to either: use a ref and an uncontrolled component or to support null when you save the value.  I think that zero isn't a valid value anyway because Plottr would better represent zero by disabling backup.
3. You can input a negative value(!)  I don't time travel, so I'm not sure what this means, but if you've found a way to reach into the future and grab new versions of books, then I want to collaborate with you on the nobel-prize-winning science! XD

Quiescent reviewed (commented):
This looks good, but I'd like to figure out what we're doing about the actual backup functionality before it lands on staging.

Quiescent committed:
Add a test file and fixtures for testing backup

Quiescent committed:
Add placeholder functions for getting backup files

Quiescent committed:
Add base backup folder to backup function skeletons

Quiescent committed:
Add tests for the backup helper functions

Quiescent committed:
Fix the directory to look for backup test files and folders

Quiescent committed:
Get the function to read anything from the backup folder

Quiescent committed:
Add skeleton of function to test whether a file is a Plottr file

Quiescent committed:
Add tests for function to test whether a file is a backup file

Quiescent committed:
Check that backup files have the right extension

Quiescent committed:
Check that plottr files are inside of backup directories

Quiescent committed:
Don't accept files in folders inside of folders as backups

Quiescent committed:
Exclude non-backup files from file listing

Quiescent committed:
Filter folders and files according to our convention for backups

Quiescent committed:
Update the current node to use from `nvm`

Quiescent committed:
Return the result of searching for backup files

Quiescent committed:
Allow backup folder days in the thirties range

Quiescent committed:
Check that the date in the thirties range is found

Quiescent committed:
Fix the regex to test for backup folders

Quiescent committed:
Fix Tests etc to filter sub files which are backups

Quiescent committed:
Implement sorting by backup file recency

Quiescent committed:
Make the start session regex more specific

Quiescent committed:
Add utility functions for date comparison with n many days ago

Quiescent committed:
Implement backup strategy

Quiescent committed:
Remove the empty directories when we're done

cameronsutter commented on src/common/utils/backup.js:
```diff
@@ -57,3 +65,112 @@ export function ensureBackupFullPath() {
+    case 'days': {
+      const anchorDate = nDaysAgo(amount)
+      const filesToDelete = files.filter((file) => !fileIsSoonerThan(anchorDate, file))
+      console.warn('`Removing old backups: ${filesToRemove}')
```
let's make this a `log.warn` from `electron-log` so it shows up in the computer's log (and error reports)

cameronsutter commented on src/common/utils/backup.js:
```diff
@@ -57,3 +65,112 @@ export function ensureBackupFullPath() {
+      const filesToDelete = files.filter((file) => !fileIsSoonerThan(anchorDate, file))
+      console.warn('`Removing old backups: ${filesToRemove}')
+      filesToDelete.forEach((file) => {
+        fs.unlinkSync(path.join(BACKUP_BASE_PATH, file))
```
there's actually a function in electron that will send these files to the trash instead of deleting them

`shell.moveItemToTrash`

(in the newest version of electron it's: https://www.electronjs.org/docs/api/shell#shelltrashitempath)

cameronsutter commented on src/common/utils/backup.js:
```diff
@@ -57,3 +65,112 @@ export function ensureBackupFullPath() {
+    }
+    case 'number': {
+      const filesToDelete = files.slice(0, files.length - amount)
+      console.warn(`Removing old backups: ${filesToDelete}`)
```
ditto

cameronsutter commented on src/common/utils/backup.js:
```diff
@@ -57,3 +65,112 @@ export function ensureBackupFullPath() {
+      const filesToDelete = files.slice(0, files.length - amount)
+      console.warn(`Removing old backups: ${filesToDelete}`)
+      filesToDelete.forEach((file) => {
+        fs.unlinkSync(path.join(BACKUP_BASE_PATH, file))
```
ditto

cameronsutter reviewed (commented):
I just had a thought. We should probably make some sort of infinite number possible, and make that the default when this gets released. Otherwise people who have been counting on the infinite backups currently are going to be stung when it erases them after 30 days … but that does bring up something. What is going to happen when this gets released?

Quiescent commented:
> I just had a thought. We should probably make some sort of infinite number possible, and make that the default when this gets released. Otherwise people who have been counting on the infinite backups currently are going to be stung when it erases them after 30 days … but that does bring up something. What is going to happen when this gets released?

Yup.  That's a great point.  I'll add a default of "Never Delete Backups".

So the way that things are at the moment (without the new "Never Delete.." feature) is that the default behaviour and amount are used.
Backups that are older than 30 days are removed when the user opens a file.

Quiescent commented:
Thanks for the review, @cameronsutter.
Here's a summary of what I'm going to fix up:
 - minor details re. Electron,
 - infini-backup, and
 - making the whole thing async.

Quiescent committed:
Add infini backup and default to it

Quiescent committed:
Use electron-log for logging

In other places I would inject the logger, but this seems to be
specific to the desktop version so I would say that it's fine here.

Quiescent committed:
Use moveItemToTrash instead of unlinking it

Quiescent committed:
Support the never-delete strategy during backup

Quiescent committed:
Don't touch disk during backup cleaning if user chose never-delete

Quiescent committed:
Change `backupFolders` into an async function

Quiescent committed:
Turn `backupFiles` into an async function

Quiescent committed:
Turn `deleteOldBackups` into an async function

Quiescent committed:
Add logging to the function to delete old backups

Quiescent committed:
Comment for posterity to indicate deletion isn't in control flow

Quiescent committed:
Fix import of promise-api for fs in Electron

Quiescent committed:
Fix files to sort argument

Quiescent committed:
Fix date arithmetic call

Quiescent committed:
Fix the library that rmdir comes from

Quiescent committed:
Don't fix the stack frames when there are none

Quiescent committed:
Improve logging

Quiescent committed:
Use promise API for rmdir

jeanawhou commented:
Looks good to me
